rust/lance/src/dataset/transaction.rs

```diff
@@ -479,52 +547,68 @@ impl Transaction {
+                    {
+                        NotCompatible
+                    } else {
+                        Compatible
+                    }
                 }
-                _ => false,
+                _ => Compatible,
             },
             // Merge changes the schema, but preserves row ids, so the only operations
             // it's compatible with is CreateIndex, ReserveFragments, SetMetadata and DeleteMetadata.
-            Operation::Merge { .. } => !matches!(
-                &other.operation,
+            Operation::Merge { .. } => match &other.operation {
                 Operation::CreateIndex { .. }
-                    | Operation::ReserveFragments { .. }
-                    | Operation::UpdateConfig { .. }
-            ),
+                | Operation::ReserveFragments { .. }
+                | Operation::UpdateConfig { .. } => Compatible,
+                Operation::Update { .. }
+                | Operation::Append { .. }
+                | Operation::Delete { .. }
+                | Operation::Rewrite { .. }
+                | Operation::Merge { .. } => Retryable,
+                _ => NotCompatible,
+            },
             Operation::Project { .. } => match &other.operation {
                 // Project is compatible with anything that doesn't change the schema
-                Operation::CreateIndex { .. } => false,
-                Operation::Overwrite { .. } => false,
-                Operation::UpdateConfig { .. } => false,
-                _ => true,
+                Operation::Append { .. }
+                | Operation::Update { .. }
+                | Operation::Delete { .. }
+                | Operation::UpdateConfig { .. }
+                | Operation::CreateIndex { .. }
+                | Operation::DataReplacement { .. } => Compatible,
+                _ => NotCompatible,
             },
             Operation::DataReplacement { .. } => match &other.operation {
                 Operation::Append { .. }
                 | Operation::Delete { .. }
                 | Operation::Update { .. }
                 | Operation::Merge { .. }
-                | Operation::UpdateConfig { .. } => false,
+                | Operation::UpdateConfig { .. } => Compatible,
                 Operation::CreateIndex { .. } => {
                     // TODO(rmeng): check that the new indices isn't on the column being replaced
-                    true
+                    NotCompatible
                 }
                 Operation::Rewrite { .. } => {
                     // TODO(rmeng): check that the fragments being replaced are not part of the groups
-                    true
+                    NotCompatible
                 }
                 Operation::DataReplacement { .. } => {
                     // TODO(rmeng): check cell conflicts
-                    true
+                    NotCompatible
                 }
-                _ => true,
+                _ => NotCompatible,
```

Comment:
These are all `NotCompatible` because this is still kind of half-finished right?  It seems `DataReplacement` could be retried in many cases?

Reply:
Yeah I figured I'll rebase https://github.com/lancedb/lance/pull/3631 on this and finish it there.